### PR TITLE
Allow for detaching serial device driver from TTY.

### PR DESCRIPTION
**Pull request by j-piecuch**
When the PTY master device file is closed, the corresponding slave device should disappear.
This PR implements the disappearing part.

### CODE REVIEW

File: sys/kern/tty.c [hunk 1]
```diff
@@ -902,6 +962,54 @@ static int tty_ioctl(file_t *f, u_long cmd, void *data) {
   }
 }
 
+static void tty_hangup(tty_t *tty) {
+  assert(mtx_owned(&tty->t_lock));
+
+  if (!tty_opened(tty) || (tty->t_lflag & CLOCAL))
+    return;
```
Comment (wiklam) [inline sys/kern/tty.c:969]: Could you provide a description why we check CLOCAL flag?

File: sys/kern/tty.c [hunk 2]
```diff
@@ -179,6 +179,29 @@ tty_t *tty_alloc(void) {
   return tty;
 }
 
+void tty_free(tty_t *tty) {
+  assert(!tty_opened(tty));
+  assert(tty_detached(tty));
+  kfree(M_DEV, tty->t_line.ln_buf);
```
Comment (wiklam) [inline sys/kern/tty.c:185]: It is not implemented yet, but `mtx_destroy` is a good idea.

File: sys/kern/tty.c [hunk 3]
```diff
@@ -902,6 +962,54 @@ static int tty_ioctl(file_t *f, u_long cmd, void *data) {
...
+  if (!tty_opened(tty) || (tty->t_lflag & CLOCAL))
+    return;
+
+  session_t *s = tty->t_session;
+
+  /* Send SIGHUP to session leader.
+   * NOTE: access to s_leader is safe here,
+   * even though we're not holding all_proc_mtx. See session_leave(). */
+  if (s && s->s_leader) {
+    proc_t *leader = s->s_leader;
```
Comment (franciscozdo) [inline sys/kern/tty.c:977]: I don't get it. Why `s->s_leader` cannot became `NULL` between this lines? Does only session leader can call this function? If not i think it is possible that this function is triggered by some process and `session_leave()` is called in session leader in the same time.

**Review by franciscozdo — APPROVED** [review]
Looks good, but I have one question about locking.

File: sys/kern/tty.c (hunk 3 above)
Comment (franciscozdo) [reply sys/kern/tty.c:977]: Okay, i missed one change. We are holding `tty->t_lock` here and in `session_leave()`. Now I see it.

File: sys/kern/tty.c (hunk 2 above)
Comment (j-piecuch) [reply sys/kern/tty.c:190]: Done.

File: sys/kern/tty.c (hunk 1 above)
Comment (j-piecuch) [reply sys/kern/tty.c:975]: Done.

**Review by wiklam — APPROVED** [review]
LGTM

**Review by cahirwpz — APPROVED** [review]
LGTM

**Pull request merged**